pkg/threescale/threescale.go

```diff
@@ -115,68 +117,95 @@ func (s *Threescale) extractProxyConf(cfg *config.Params, c *sysC.ThreeScaleClie
 	return pce, proxyConfErr
 }
 
-// isAuthorized returns code 0 if authorization is successful
-// based on grpc return codes https://github.com/grpc/grpc-go/blob/master/codes/codes.go
-func (s *Threescale) isAuthorized(cfg *config.Params, request authorization.InstanceMsg, proxyConf sysC.ProxyConfigElement, client *backendC.ThreeScaleClient) (rpc.Status, error) {
+// isAuthorized - is responsible for parsing the incoming request and determining if it is valid, building out the request to be sent
+// to 3scale and parsing the response. Returns code 0 if authorization is successful based on
+// grpc return codes https://github.com/grpc/grpc-go/blob/master/codes/codes.go
+func (s *Threescale) isAuthorized(svcID string, request authorization.InstanceMsg, proxyConf sysC.ProxyConfig, client *backendC.ThreeScaleClient) rpc.Status {
+	var (
+		// Application ID authentication pattern - App Key is optional when using this authn
+		appID, appKey string
+
+		// Application Key auth pattern
+		userKey string
+
+		// Function to be called when authn patter has been determined
+		authRep authRepFn
+	)
+
 	if request.Action.Path == "" {
-		return status.WithInvalidArgument("missing request path"), nil
+		request.Action.Path = "/"
 	}
 
-	userKey, err := parseActionPath(request.Action.Path)
-	if err != nil {
-		return status.WithPermissionDenied(err.Error()), nil
+	metrics := generateMetrics(request.Action.Path, request.Action.Method, proxyConf)
+	if len(metrics) == 0 {
+		return status.WithPermissionDenied(fmt.Sprintf("no matching mapping rule for request with method %s and path %s",
+			request.Action.Method, request.Action.Path))
+	}
+
+	authRepRequest := authRepRequest{
+		svcID: svcID,
+		auth: backendC.TokenAuth{
+			Type:  proxyConf.Content.BackendAuthenticationType,
+			Value: proxyConf.Content.BackendAuthenticationValue,
+		}}
+
+	if request.Subject != nil {
+		appID = request.Subject.Properties["app_id"].GetStringValue()
+		appKey = request.Subject.Properties["app_key"].GetStringValue()
+
+		userKey = request.Subject.User
 	}
 
-	return s.doAuthRep(cfg.ServiceId, userKey, request, proxyConf.ProxyConfig, client)
+	if appID == "" && userKey == "" {
+		return status.WithPermissionDenied(unauthenticatedErr)
+	}
+
+	// if app_key is provided, enforce App Id authentication pattern
+	if appKey != "" && appID == "" {
+		return status.WithPermissionDenied(authMismatchErr)
+
+	} else if appID != "" {
+		authRepRequest.params = backendC.NewAuthRepParamsAppID(appKey, "", "", metrics, nil)
+		authRep = client.AuthRepAppID
+
+	} else {
+		authRepRequest.authKey = request.Subject.User
+		authRepRequest.params = backendC.NewAuthRepParamsUserKey("", "", metrics, nil)
+		authRep = client.AuthRepUserKey
+	}
```

Comment:
so, for all of this, I have commented about the order elsewhere, but I'd prefer to:
1. Use `user_key` if present, forget about the rest.
2. Otherwise use `app_id` if present. If `app_key` is present, also use it with `app_id`.
3. Otherwise error out.

Reply:
Ok I'm fine with that.

Reply:
Done